feat(components): [skeleton] `throttle` supports values ​​as object

chenweiyi opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### Solving the problem

When using the `throttle` property of skeleton in the first screen scene, it does not take effect because the current `throttle` property only deals with the *delay display* issue and does not solve the *delay hide* issue. Therefore, I have extended this attribute value to support configuring the *delay hide* attribute.

Now throttle support config object like `{ trailing: 500 }`

### Relate issue
Through config `{ trailing: xxx, initVal: true }`   fix #4804  

### Rendering

![image](https://github.com/user-attachments/assets/3752e21c-739c-4a82-b24f-2e147a7caa62)

chenweiyi committed:
fix(components): [skeleton] `throttle` property not working

chenweiyi committed:
fix(components): [skeleton] `throttle` property not working

chenweiyi committed:
fix: lint fix

chenweiyi committed:
fix: lint fix

chenweiyi committed:
feat: add func & doc & test

chenweiyi committed:
feat: add func & doc & test

chenweiyi commented:
/label Components:[skeleton]

chenweiyi committed:
feat: remove test modify

chenweiyi committed:
feat: remove test modify

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'element-plus:dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
feat: increase document examples, improve document descriptions

chenweiyi committed:
feat: increase document examples, improve document descriptions

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' of github.com:chenweiyi/element-plus into dev

chenweiyi committed:
Merge branch 'dev' of github.com:element-plus/element-plus into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'element-plus:dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

btea commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,29 +1,60 @@
+    } else {
+      throttled.value = false
+    }
+  }
```
There is a lot of overlap between these two functions, and perhaps they can be combined into one function.

Some type determination operations can directly reference internal encapsulated tool functions.
https://github.com/element-plus/element-plus/blob/72dd11aac271e788da45e0e92b3ad3dd343c1601/packages/utils/types.ts#L18

chenweiyi replied:
I'm not quite clear about what you mean. The `use-throttle-render/index.ts` mainly exposes a throttle hook, and `utlils/types.ts` contains mostly pure function logic. These two can't be simply merged into one, can they? Could you please elaborate a bit more, and then I can see what I might be able to do?

btea replied:
Sorry, it looks like my description was a bit confusing.

 That some of the code logic of function `leadingDispatch` and function `trailingDispatch` is the same. These two functions can be merged into one function by passing a Boolean value to distinguish the different logics.

The second sentence describes that code like `typeof throttle === 'number'` can be processed by referencing the internal encapsulated function and replaced with `isNumber(throttle)`.

chenweiyi replied:
Thank you for your patient explanation, I have made a revision, please review it again.

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
feat: 重构`useThrottleRender`钩子以提高代码可读性和效率

- 简化了对`throttle`参数的判断逻辑，通过`isNumber`函数判断是否为数字
- 将`leadingDispatch`和`trailingDispatch`函数合并为`dispatcher`函数，根据传入的类型判断执行逻辑
- 优化了`watch`回调函数，使用`dispatcher`函数替代重复的判断逻辑

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
feat: 写法优化

chenweiyi committed:
Merge branch 'dev' into dev

btea commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,31 +1,58 @@
+      }
+    } else {
+      if (
+        typeof throttle === 'object' &&
```
```suggestion
        isObject(throttle) &&
```

chenweiyi committed:
feat: 引入`isObject`函数替代原有的`typeof throttle === 'object'`判断方式

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

btea commented on docs/en-US/component/skeleton.md:
```diff
@@ -77,7 +77,7 @@ skeleton/rendering-with-data
 ## Avoiding rendering bouncing.
 
-Sometimes API responds very quickly, when that happens, the skeleton just gets rendered to the DOM then it needs to switch back to real DOM, that causes the sudden flashy. To avoid such thing, you can use the `throttle` attribute.
+Sometimes API responds very quickly, when that happens, the skeleton just gets rendered to the DOM then it needs to switch back to real DOM, that causes the sudden flashy. To avoid such thing, you can use the `throttle` attribute. The `throttle` attribute supports two values: `number` and `object`. When passing a `number`, it is equivalent to `{leading: xxx}`, controlling the throttling of the sk ...
```
The content here may be created as a separate subheading, and the supported version number can be conveniently marked.

chenweiyi replied:
ok

btea commented:
Right now, the content appears and disappears a bit abruptly. Maybe we can add a transition effect to make it look softer.
WDYT？@chenweiyi

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
feat: 优化骨架屏文档结构和示例

chenweiyi commented:
> Right now, the content appears and disappears a bit abruptly. Maybe we can add a transition effect to make it look softer. WDYT？@chenweiyi

I create two sperate section for describle the new function. Please take a look at.

chenweiyi committed:
feat: 完善文字描述和修改对应的文件名

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

btea commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,31 +1,53 @@
+) => {
   if (throttle === 0) return loading
-  const throttled = ref(false)
+  const initVal = isObject(throttle) ? throttle.initVal ?? false : false
```
It seems that there is no need for false to fall back here.
```suggestion
  const initVal = isObject(throttle) ? throttle.initVal : false
```

chenweiyi replied:
If remove `throttle.initVal ?? false` , `useThrottleRender` will return `ref<boolean | undefined>`, But I think `useThrottleRender` of return `ref<boolean>` express more clear meaning.

![image](https://github.com/user-attachments/assets/d79f7af9-24e0-4123-92e9-5e05e027b9bc)

![image](https://github.com/user-attachments/assets/8d2e0fc2-c3e3-42c1-a63d-09d25f328d1b)

chenweiyi replied:
I have thought about modifying it to this `const initVal = isObject(throttle) && Boolean(throttle.initVal)`

btea commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,31 +1,53 @@
+        }
+      }
+    } else {
+      if (isObject(throttle) && throttle?.trailing && throttle.trailing > 0) {
```
```suggestion
      if (isObject(throttle)) {
```

chenweiyi replied:
I think `isObject(throttle) && throttle?.trailing && throttle.trailing > 0` can be changed to `Object(throttle) && throttle?.trailing`. The judgment of `throttle.trailing > 0` is essential, otherwise the parameter `timer` in `dispatchThrottling` may throw an error in `setTimeout`.

btea replied:
The second parameter of `setTimeout` will not throw an error even if it accepts `undefined` or a negative number. Of course, you can make a more accurate judgment, but we need to make a backup. When the conditions are not met, perhaps we should give a default value of `0`.

chenweiyi replied:
I have modified the `dispatcher` method again：
```
const dispatcher = (type: 'leading' | 'trailing') => {
    if (type === 'leading') {
      if (isNumber(throttle)) {
        dispatchThrottling(throttle)
      } else {
        dispatchThrottling(throttle.leading)
      }
    } else {
      if (isObject(throttle)) {
        dispatchThrottling(throttle.trailing)
      } else {
        throttled.value = false
      }
    }
  }
```
Then `dispatchThrottling` was modified, adding handling for the `undefined` case as well ：
```
const dispatchThrottling = (timer: number | undefined) => {
    if (isUndefined(timer)) {
      throttled.value = loading.value
      return
    }
    if (timeoutHandle) {
      clearTimeout(timeoutHandle)
    }
    timeoutHandle = setTimeout(() => {
      throttled.value = loading.value
    }, timer)
  }
```

btea commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,31 +1,53 @@
+      if (isNumber(throttle)) {
+        dispatchThrottling(throttle)
+      } else {
+        if (throttle?.leading && throttle.leading > 0) {
```
The judgment here can be removed, otherwise, there will be problems when it is set to 0.

chenweiyi replied:
Here `isNumber(throttle)` is for compatibility with the previous logic. `throttle?.leading && throttle.leading > 0` is here to determine that the `throttle` passed in by the user is an object, but the `leading` attribute is not passed. Here I think the judgment of `throttle.leading > 0` can also be removed.

btea replied:
Yes, that's what I meant.

chenweiyi replied:
I have thought about modifying it to this `const initVal = isObject(throttle) && Boolean(throttle.initVal)`

btea commented on docs/en-US/component/skeleton.md:
```diff
@@ -79,23 +79,57 @@ skeleton/rendering-with-data
 
 :::
 
+## Initial rendering loading
```
```suggestion
## Initial rendering loading ^(2.8.7)
```

btea commented on docs/en-US/component/skeleton.md:
```diff
@@ -79,23 +79,57 @@ skeleton/rendering-with-data
+
+:::
+
+## Toggle show/hide without rending bouncing
```
```suggestion
## Toggle show/hide without rending bouncing ^(2.8.7)
```

chenweiyi committed:
Update docs/en-US/component/skeleton.md

Co-authored-by: btea <2356281422@qq.com>

chenweiyi committed:
Update docs/en-US/component/skeleton.md

Co-authored-by: btea <2356281422@qq.com>

chenweiyi committed:
feat: Optimize code writing

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

btea reviewed (approved):
Thanks!

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi commented:
> Thanks!

Thank you for reviewing my code very patiently. I am glad I can contribute.

chenweiyi committed:
Merge branch 'dev' into dev

chenweiyi committed:
Update docs/en-US/component/skeleton.md

Co-authored-by: btea <2356281422@qq.com>

chenweiyi commented on docs/en-US/component/skeleton.md:
```diff
@@ -79,23 +79,57 @@ skeleton/rendering-with-data
 
 :::
 
+## Initial rendering loading ^(2.8.7)
```
Ok, Thank you!

btea commented on docs/en-US/component/skeleton.md:
```diff
@@ -79,23 +79,57 @@ skeleton/rendering-with-data
+
+:::
+
+## Toggle show/hide without rending bouncing ^(2.8.7)
```
```suggestion
## Toggle show/hide without rending bouncing ^(2.8.8)
```

btea committed:
Update docs/en-US/component/skeleton.md

chenweiyi committed:
Merge branch 'dev' into dev

tolking commented on packages/hooks/use-throttle-render/index.ts:
```diff
@@ -1,31 +1,55 @@
 import { onMounted, ref, watch } from 'vue'
+import { isNumber, isObject, isUndefined } from '@element-plus/utils'
 
+import type { SkeletonThrottle } from '@element-plus/components'
```
Avoid import components content in hooks. It is recommended to export the `SkeletonThrottle` type here.

tolking replied:
This issue does not seem to be resolved

Suggest moving the type declaration of `SkeletonThrottle` to here

chenweiyi replied:
Okay, I understand what you mean, I will also change the `SkeletonThrottle` field to `ThrottleType` to increase generality.

tolking commented on docs/en-US/component/skeleton.md:
```diff
@@ -79,23 +79,57 @@ skeleton/rendering-with-data
 
+:::tip
+
+The `throttle` attribute supports two values: `number` and `object`. When passing a `number`, it is equivalent to `{leading: xxx}`, controlling the throttling of the skeleton screen display. Of course, you can also control the throttling of the skeleton screen disappearance by passing `{trailing: xxx}`
```
It would be better to add a version hint here. 

eg. Since ^(2.8.8), ...

tolking reviewed (commented):
Could you add some test cases for useThrottleRender?

chenweiyi committed:
feat: modify doc

chenweiyi committed:
feat: md

chenweiyi committed:
feat: 补充 useThrottleRender 钩子的测试用例

chenweiyi commented:
> Could you add some test cases for useThrottleRender?

I added 2 test cases, thank you for your review.

chenweiyi committed:
feat: 将 SkeletonThrottle 类型移动到hook中， 重命名为 ThrottleType 以提高通用性

chenweiyi committed:
Merge branch 'dev' into dev
